propagation.go

```diff
@@ -0,0 +1,103 @@
+package opentracing
+
+///////////////////////////////////////////////////////////////////////////////
+// CORE PROPAGATION INTERFACES:
+///////////////////////////////////////////////////////////////////////////////
+
+// PropagationInjector is responsible for injecting Span instances in a manner
+// suitable for propagation via a format-specific "carrier" object. Typically
+// the injection will take place across an RPC boundary, but message queues and
+// other IPC mechanisms are also reasonable places to use a
+// PropagationInjector.
+//
+// See PropagationExtractor and Span.PropagationInjectorForFormat.
+type PropagationInjector interface {
+	InjectSpan(span Span, carrier interface{})
+}
+
+// PropagationExtractor is responsible for extracting Span instances from an
+// format-specific "carrier" object. Typically the extraction will take place
+// on the server side of an RPC boundary, but message queues and other IPC
+// mechanisms are also reasonable places to use a PropagationExtractor.
+//
+// See PropagationInjector and Tracer.PropagationExtractorForFormat.
+type PropagationExtractor interface {
+	ExtractSpan(operationName string, carrier interface{}) (Span, error)
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// BUILTIN PROPAGATION FORMATS:
+///////////////////////////////////////////////////////////////////////////////
+
+// BuiltinPropagationFormat is the shared type for builtin OpenTracing in-band
+// propagation formats.
+type BuiltinPropagationFormat int
+
+const (
+	// PROPAGATION_FORMAT_SPLIT_BINARY encodes the Span in a BinaryCarrier
+	// instance.
+	//
+	// The `carrier` for injection and extraction must be a `*BinaryCarrier`
+	// instance.
+	PROPAGATION_FORMAT_SPLIT_BINARY BuiltinPropagationFormat = iota
+
+	// PROPAGATION_FORMAT_SPLIT_BINARY encodes the Span in a TextCarrier
+	// instance.
+	//
+	// The `carrier` for injection and extraction must be a `*TextCarrier`
+	// instance.
+	PROPAGATION_FORMAT_SPLIT_TEXT
+
+	// PROPAGATION_FORMAT_GO_HTTP_HEADER encodes the Span into a Go http.Header
+	// instance (both the tracer state and any Trace Attributes).
+	//
+	// The `carrier` for both injection and extraction must be an http.Header
+	// instance.
+	PROPAGATION_FORMAT_GO_HTTP_HEADER
+)
+
+// TextCarrier breaks a propagated Span into two pieces.
+//
+// The Span is separated in this way for a variety of reasons; the most
+// important is to give OpenTracing users a portable way to opt out of Trace
+// Attribute propagation entirely if they deem it a stability risk.
+type TextCarrier struct {
```

Comment:
@bensigelman why not use an interface that does not require intermediate map allocation?

```go
type TextInjector interface {
    AddTracerStateField(key, value string)
    AddTraceAttribute(key, value string)
}
// util function to construct TextInjector from maps
func NewMapBasedTextInjector(tracerState map[string]string, traceAttributes map[string]string) TextInjector {}

// util function to construct TextInjector for http.Header
func NewHttpHeaderTextInjector(header *http.Header) TextInjector {}
```

Note that this allows varying the argument to the Injector's inject method without varying the format.

The extractor part is a bit tricker, as it needs direct by-key access for tracer state, and possibly iterator over all keys for trace attributes.

On second thought, this doesn't have to *replace* TextCarrier, could be another standard format (better performance, less copying).